propolis/src/block/mod.rs

```diff
@@ -140,38 +205,83 @@ pub trait Device: Send + Sync + 'static {
 }
 
 pub trait Backend: Send + Sync + 'static {
-    fn attach(&self, dev: Arc<dyn Device>, disp: &Dispatcher);
+    fn attach(
+        &self,
+        dev: Arc<dyn Device>,
+        disp: &Dispatcher,
+    ) -> std::io::Result<()>;
     fn info(&self) -> DeviceInfo;
 }
 
 pub type NotifierFn = dyn Fn(&dyn Device, &DispCtx) + Send + Sync + 'static;
 
+/// Notifier help by every block device used to indicate
+/// to the corresponding block backends about I/O requests.
 pub struct Notifier {
+    /// Flag used to coalesce request notifications from the block device.
+    ///
+    /// If set, we've previously been notified but the backend has yet to
+    /// respond to the latest notification.
     armed: AtomicBool,
+
+    /// The backend specific notification handler
     notifier: Mutex<Option<Box<NotifierFn>>>,
+
+    /// Whether or not we should service I/O requests
+    paused: AtomicBool,
+
+    /// Count of how many outstanding I/O requests there currently are
+    outstanding_reqs: Arc<AtomicU64>,
+
+    /// Notifier to indicate all outstanding requests have been completed
+    outstanding_notifier: Arc<Mutex<Option<Arc<Notify>>>>,
 }
+
 impl Notifier {
     pub fn new() -> Self {
-        Self { armed: AtomicBool::new(false), notifier: Mutex::new(None) }
+        Self {
+            armed: AtomicBool::new(false),
+            notifier: Mutex::new(None),
+            paused: AtomicBool::new(false),
+            outstanding_reqs: Arc::new(AtomicU64::new(0)),
+            outstanding_notifier: Arc::new(Mutex::new(None)),
+        }
     }
     pub fn next_arming(
         &self,
         nextf: impl Fn() -> Option<Request>,
     ) -> Option<Request> {
+        if self.paused.load(Ordering::Acquire) {
+            return None;
+        }
+
         self.armed.store(false, Ordering::Release);
-        let res = nextf();
-        if res.is_some() {
-            // Since a result was successfully retrieved, no need to rearm the
-            // notification trigger.
-            return res;
+        if let Some(mut req) = nextf() {
+            // Update outstanding count and update `Request` to track it
+            self.outstanding_reqs.fetch_add(1, Ordering::Relaxed);
```

Comment:
Is `Relaxed` ordering correct here?

Reply:
I believe so, we just want to atomicity of the increment here and do nothing with the old value. There's no other memory location we're trying to synchronize with if a thread gets to this point before it returns from this method.